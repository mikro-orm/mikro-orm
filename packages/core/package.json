{
  "name": "@mikro-orm/core",
  "version": "6.1.12",
  "description": "TypeScript ORM for Node.js based on Data Mapper, Unit of Work and Identity Map patterns. Supports MongoDB, MySQL, PostgreSQL and SQLite databases as well as usage with vanilla JavaScript.",
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "typings": "dist/index.d.ts",
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "import": {
        "types": "./dist/index.d.ts",
        "default": "./dist/index.mjs"
      },
      "require": "./dist/index.js"
    }
  },
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com/mikro-orm/mikro-orm.git"
  },
  "funding": "https://github.com/sponsors/b4nan",
  "keywords": [
    "orm",
    "mongo",
    "mongodb",
    "mysql",
    "mariadb",
    "postgresql",
    "sqlite",
    "sqlite3",
    "ts",
    "typescript",
    "js",
    "javascript",
    "entity",
    "ddd",
    "mikro-orm",
    "unit-of-work",
    "data-mapper",
    "identity-map"
  ],
  "author": "Martin AdÃ¡mek",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/mikro-orm/mikro-orm/issues"
  },
  "homepage": "https://mikro-orm.io",
  "engines": {
    "node": ">= 18.12.0"
  },
  "scripts": {
    "build": "yarn clean && yarn compile && yarn copy && yarn run -T gen-esm-wrapper dist/index.js dist/index.mjs",
    "clean": "yarn run -T rimraf ./dist",
    "compile": "yarn run -T tsc -p tsconfig.build.json",
    "copy": "node ../../scripts/copy.mjs"
  },
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "dataloader": "2.2.2",
    "dotenv": "16.4.5",
    "esprima": "4.0.1",
    "fs-extra": "11.2.0",
    "globby": "11.1.0",
    "mikro-orm": "6.1.12",
    "reflect-metadata": "0.2.2"
  }
}
