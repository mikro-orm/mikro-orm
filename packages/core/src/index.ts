/**
 * @packageDocumentation
 * @module core
 */
/* istanbul ignore file */
export {
  Constructor, Dictionary, PrimaryKeyType, PrimaryKeyProp, Primary, IPrimaryKey, FilterQuery, IWrappedEntity, EntityName, EntityData, Highlighter,
  AnyEntity, EntityProperty, EntityMetadata, QBFilterQuery, PopulateOptions, Populate, Loaded, New, LoadedReference, LoadedCollection,
  GetRepository, EntityRepositoryType, MigrationObject, DeepPartial, PrimaryProperty, Cast, IsUnknown, EntityDictionary, EntityDTO, PlainObject,
} from './typings';
export * from './enums';
export * from './errors';
export * from './exceptions';
export * from './MikroORM';
export * from './entity';
export * from './events';
export * from './EntityManager';
export * from './unit-of-work';
export * from './utils';
export * from './hydration';
export * from './drivers';
export * from './connections';
export * from './platforms';
export * from './types';
export * from './naming-strategy';
export * from './metadata';
export * from './cache';
export * from './decorators';
