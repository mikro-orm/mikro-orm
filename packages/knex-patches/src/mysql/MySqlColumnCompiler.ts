// @ts-ignore
import BaseMySqlColumnCompiler from 'knex/lib/dialects/mysql/schema/mysql-columncompiler';

export class MySqlColumnCompiler extends BaseMySqlColumnCompiler {

  // we need the old behaviour to be able to add auto_increment to a column that is already PK
  increments(this: any, options = { primaryKey: true }) {
    return 'int unsigned not null auto_increment' + (this.tableCompiler._canBeAddPrimaryKey(options) ? ' primary key' : '');
  }

  /* istanbul ignore next */
  bigincrements(this: any, options = { primaryKey: true }) {
    return 'bigint unsigned not null auto_increment' + (this.tableCompiler._canBeAddPrimaryKey(options) ? ' primary key' : '');
  }

}
