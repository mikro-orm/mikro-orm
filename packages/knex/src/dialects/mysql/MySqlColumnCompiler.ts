// @ts-ignore
import BaseMySqlColumnCompiler from 'knex/lib/dialects/mysql/schema/mysql-columncompiler';

export class MySqlColumnCompiler extends BaseMySqlColumnCompiler {

  // we need the old behaviour to be able to add auto_increment to a column that is already PK
  increments(this: any, options = { primaryKey: true, unsigned: true }) {
    const type = options.unsigned ? 'int unsigned' : 'int';
    return `${type} not null auto_increment` + (this.tableCompiler._canBeAddPrimaryKey(options) ? ' primary key' : '');
  }

  /* istanbul ignore next */
  bigincrements(this: any, options = { primaryKey: true, unsigned: true }) {
    const type = options.unsigned ? 'bigint unsigned' : 'bigint';
    return `${type} not null auto_increment` + (this.tableCompiler._canBeAddPrimaryKey(options) ? ' primary key' : '');
  }

}
