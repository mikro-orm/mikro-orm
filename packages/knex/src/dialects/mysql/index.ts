export * from './MariaDbKnexDialect';
export * from './MySqlKnexDialect';
export * from './MySqlConnection';
export * from './MySqlExceptionConverter';
export * from './MySqlSchemaHelper';
export * from './MySqlPlatform';
