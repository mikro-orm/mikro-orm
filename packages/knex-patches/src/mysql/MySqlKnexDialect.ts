import { MySqlQueryCompiler } from './MySqlQueryCompiler';
import { MySqlColumnCompiler } from './MySqlColumnCompiler';
import { MonkeyPatchable } from '../MonkeyPatchable';

export class MySqlKnexDialect extends MonkeyPatchable.MySqlDialect {

  queryCompiler() {
    // eslint-disable-next-line prefer-rest-params
    return new (MySqlQueryCompiler as any)(this, ...arguments);
  }

  columnCompiler() {
    // eslint-disable-next-line prefer-rest-params
    return new (MySqlColumnCompiler as any)(this, ...arguments);
  }

}
