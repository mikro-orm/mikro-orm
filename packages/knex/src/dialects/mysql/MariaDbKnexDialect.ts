import { MySqlKnexDialect } from './MySqlKnexDialect';

export class MariaDbKnexDialect extends MySqlKnexDialect {

  get driverName() {
    return 'mariadb';
  }

  _driver() {
    return require('mariadb/callback');
  }

  validateConnection(connection: any) {
    return connection.isValid();
  }

}
