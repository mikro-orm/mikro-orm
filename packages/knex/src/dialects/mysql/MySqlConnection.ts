import type { Knex } from 'knex';
import { MySqlKnexDialect } from './MySqlKnexDialect';
import { AbstractSqlConnection } from '../../AbstractSqlConnection';

export class MySqlConnection extends AbstractSqlConnection {

  override createKnex() {
    this.client = this.createKnexClient(MySqlKnexDialect as any);
    this.connected = true;
  }

  getDefaultClientUrl(): string {
    return 'mysql://root@127.0.0.1:3306';
  }

  override getConnectionOptions(): Knex.MySqlConnectionConfig {
    const ret = super.getConnectionOptions() as Knex.MySqlConnectionConfig;

    if (this.config.get('multipleStatements')) {
      ret.multipleStatements = this.config.get('multipleStatements');
    }

    if (this.config.get('forceUtcTimezone')) {
      ret.timezone = 'Z';
    }

    if (this.config.get('timezone')) {
      ret.timezone = this.config.get('timezone');
    }

    ret.supportBigNumbers = true;
    ret.dateStrings = true;

    return ret;
  }

  protected transformRawResult<T>(res: any, method: 'all' | 'get' | 'run'): T {
    if (method === 'run' && ['OkPacket', 'ResultSetHeader'].includes(res[0].constructor.name)) {
      return {
        insertId: res[0].insertId,
        affectedRows: res[0].affectedRows,
        rows: [],
      } as unknown as T;
    }

    if (method === 'get') {
      return res[0][0];
    }

    return res[0];
  }

}
