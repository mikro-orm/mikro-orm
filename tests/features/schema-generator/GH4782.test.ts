import { MikroORM, sql } from '@mikro-orm/mysql';
import { Entity, PrimaryKey, Property } from '@mikro-orm/core';

@Entity({ tableName: 'user' })
class User0 {

  @PrimaryKey()
  id!: number;

}

@Entity({ tableName: 'user' })
class User1 {

  @PrimaryKey()
  id!: number;

  @Property({ default: sql.now(3), columnType: 'timestamp(3)' })
  bar!: Date;

}

@Entity({ tableName: 'user' })
class User2 {

  @PrimaryKey()
  id!: number;

  @Property({ default: sql.now(3), columnType: 'timestamp(3)' })
  bar!: Date;

}

@Entity({ tableName: 'user' })
class User3 {

  @PrimaryKey()
  id!: number;

  @Property({ default: sql.now(6), columnType: 'timestamp(6)' })
  bar!: Date;

}

@Entity({ tableName: 'user' })
class User4 {

  @PrimaryKey()
  id!: number;

}

let orm: MikroORM;

beforeAll(async () => {
  orm = await MikroORM.init({
    entities: [User0],
    dbName: 'mikro_orm_test_gh_4782',
    port: 3308,
  });
  await orm.schema.ensureDatabase();
  await orm.schema.dropSchema();
});

afterAll(() => orm.close(true));

test('4782', async () => {
  const testMigration = async (e1: any, e2: any, snap: string) => {
    if (e2) {
      orm.getMetadata().reset(e1.name);
      await orm.discoverEntity(e2);
    }

    const diff = await orm.schema.getUpdateSchemaMigrationSQL({ wrap: false });
    expect(diff).toMatchSnapshot(snap);
    await orm.schema.execute(diff.up);

    return diff.down;
  };

  const down: string[] = [];
  down.push(await testMigration(User0, undefined, '0. create schema'));
  down.push(await testMigration(User0, User1, '1. add timestamp(3) column'));
  down.push(await testMigration(User1, User2, '2. no changes'));
  down.push(await testMigration(User2, User3, '3. change to timestamp(6)'));
  down.push(await testMigration(User3, User4, '4. remove timestamp column'));

  for (const sql of down.reverse()) {
    await orm.schema.execute(sql);
  }
});
