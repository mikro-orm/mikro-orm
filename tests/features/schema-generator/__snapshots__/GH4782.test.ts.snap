// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`4782: 0. create schema 1`] = `
{
  "down": "drop table if exists \`user\`;

",
  "up": "create table \`user\` (\`id\` int unsigned not null auto_increment primary key) default character set utf8mb4 engine = InnoDB;

",
}
`;

exports[`4782: 1. add timestamp(3) column 1`] = `
{
  "down": "alter table \`user\` drop column \`bar\`;

",
  "up": "alter table \`user\` add \`bar\` timestamp(3) not null default current_timestamp(3);

",
}
`;

exports[`4782: 2. no changes 1`] = `
{
  "down": "",
  "up": "",
}
`;

exports[`4782: 3. change to timestamp(6) 1`] = `
{
  "down": "alter table \`user\` modify \`bar\` timestamp(3) not null default current_timestamp(3);

",
  "up": "alter table \`user\` modify \`bar\` timestamp(6) not null default current_timestamp(6);

",
}
`;

exports[`4782: 4. remove timestamp column 1`] = `
{
  "down": "alter table \`user\` add \`bar\` timestamp(6) not null default current_timestamp(6);

",
  "up": "alter table \`user\` drop column \`bar\`;

",
}
`;
